src/client/datascience/jupyter/liveshare/hostJupyterServer.ts

```diff
@@ -310,26 +310,36 @@ export class HostJupyterServer extends LiveShareParticipantHost(JupyterServerBas
         // Do this only if kernel information has been provided in the metadata, or the resource's interpreter is different.
         let changedKernel = false;
         if (
-            notebookMetadata?.kernelspec ||
-            notebookMetadata?.id ||
+            kernelConnection ||
             resourceInterpreter?.displayName !== launchInfo.kernelConnectionMetadata?.interpreter?.displayName
         ) {
-            const kernelInfo = await (launchInfo.connectionInfo.localLaunch
+            // TODO: Need to read and ensure this logic is correct.
+            // Basically, if we have the connection information, then there's no need to perform any search.
+            // If we don't have the kernel connection information, then we haven't performed the search
+            // Is this assumption correct, do we not search for kernel information when starting a notebook without selecting a kernel.
+            // Doesn't extension today rely on this code to select a default kernel? (mayby not)..
```

Comment:
Some personal TODOs: I need to check if we even need `getPreferredKernelForLocalConnection`
Its possible we need it, but i would like to understand it better (possible its only required for live share, if thats the case, i'll update comments).
FYI - I don't see when the `getPreferredKernelForLocalConnection` will get called for non-live share scenearios - hence the note

Reply:
This is a WIP